src/llvm/mod.rs

```diff
@@ -179,65 +178,44 @@ pub unsafe fn optimize(
         LLVMSetModuleInlineAsm2(module, ptr::null_mut(), 0);
     }
 
-    let mpm = LLVMCreatePassManager();
-    let fpm = LLVMCreateFunctionPassManagerForModule(module);
-
-    LLVMAddAnalysisPasses(tm, mpm);
-    LLVMAddAnalysisPasses(tm, fpm);
-
-    // even with -O0 and without LTO we still want to avoid linking in unused code from core etc
-    LLVMAddGlobalDCEPass(mpm);
-
-    let pmb = LLVMPassManagerBuilderCreate();
-
-    use OptLevel::*;
-    let (inline_threshold, opt) = match opt_level {
-        No | SizeMin => (0, 1), // Pretty much nothing compiles with -O0 s∫o make it an alias for -O1
-        Less => (25, 1),
-        Default => (225, 2),
-        Aggressive => (275, 3),
-        Size => (25, 0),
-    };
-    LLVMPassManagerBuilderSetOptLevel(pmb, opt);
-    LLVMPassManagerBuilderSetSizeLevel(
-        pmb,
-        match opt_level {
-            Size => 1,
-            SizeMin => 2,
-            _ => 0,
-        },
-    );
-    LLVMPassManagerBuilderUseInlinerWithThreshold(pmb, inline_threshold);
-
-    // populate the pass managers
-    LLVMPassManagerBuilderPopulateFunctionPassManager(pmb, fpm);
-    LLVMPassManagerBuilderPopulateModulePassManager(pmb, mpm);
-
     for sym in module.globals_iter() {
         internalize(sym, symbol_name(sym), export_symbols);
     }
     for sym in module.global_aliases_iter() {
         internalize(sym, symbol_name(sym), export_symbols);
     }
 
-    debug!("running function passes");
-    LLVMInitializeFunctionPassManager(fpm);
     for function in module.functions_iter() {
         let name = symbol_name(function);
         if !name.starts_with("llvm.") {
             if ignore_inline_never {
                 remove_attribute(function, "noinline");
             }
             internalize(function, name, export_symbols);
-            if LLVMIsDeclaration(function) == 0 {
-                LLVMRunFunctionPassManager(fpm, function);
-            }
         }
     }
-    LLVMFinalizeFunctionPassManager(fpm);
 
-    debug!("running module passes");
-    LLVMRunPassManager(mpm, module);
+    let options = LLVMCreatePassBuilderOptions();
+
+    // even with -O0 and without LTO we still want to avoid linking in unused code from core etc
+    let dce = CString::new("dce").unwrap();
+    debug!("running dce passe");
+    LLVMRunPasses(module, dce.as_ptr(), tm, options);
```

Comment:
I'm not sure that running this twice is the proper way to do it. Have you found out why "dce,default<whatever>" doesn't work? rustc does seem to implement extra passes that way

Reply:
I have no idea why, but it *does not* seem to work.

Reply:
It seems not possible to combine default with any others?

Reply:
I added a branch with dead code, would you like to experiment with it after this lands?

Reply:
I'd like to experiment with it before it lands, but after sleeping